Android Embedding Refactor PR31: Integrate platform views with the new embedding and the plugin shim.

matthew-carroll opened the pull request:
Android Embedding Refactor PR31: Integrate platform views with the new embedding and the plugin shim.

To test this change, I ported the WebView plugin locally to use the new Android embedding. It works as expected.

This PR adjusts where and when the `PlatformViewsController` is created and attached to an `Activity` to respect the possibility that an `Activity` does not initially exist. It does so in a way that still allows the `ShimPluginRegistry` to work as expected with existing plugins in a full-Flutter scenario.

Additional investigation and work on `PlatformViewsController` is still necessary, but that future work is not expected to require fundamental changes to the embedding APIs, so I've left it for future PRs. With this PR, we should be able to say that platform views are integrated with the new embedding, but without leaking an `Activity` and without crashing some bots.

This PR should constitute the completion of known requirements for the core embedding + plugin API (minus any future decisions about plugin registration, etc.).

mklim commented on shell/platform/android/io/flutter/embedding/android/FlutterView.java:
```diff
@@ -112,6 +116,8 @@ public void onFirstFrameRendered() {
    *   <li>{@link #renderMode} defaults to {@link RenderMode#surface}.</li>
    *   <li>{@link #transparencyMode} defaults to {@link TransparencyMode#opaque}.</li>
    * </ul>
+   * {@code FlutterView} requires an {@code Activity} instead of a generic {@code Context}
```
Should the signature be changed to actually take an Activity throughout?

xster replied:
I thought the approach we were talking about last week was that we're keeping the PlatformViewController to have an ApplicationContext constructor but just that we're adding new APIs to it, something like onAttachToActivity. And you can only do presentation.show after you onAttachToActivity. But then FlutterView isn't involved in any of this. It's just between the fragment, which does flutterPluginRegistry.getPlatformViewsController.onAttachActivity(activity), and the PlatformViewsController.

matthew-carroll replied:
@mklim I tried that but compilation failed because apparently Android `View` subclasses need to declare exact replicas of superclass constructors (probably due to XML inflation concerns). So I backed out that change and left a comment.

@xster the reason this was a problem at all is because we can't use an application context. If you ever attempt to `show()` a dialog/presentation with an application context, the app will blow up. That's why the previous commit that used the application context had to be reverted, which sparked the conversation with the Android team, etc. So now we're doing what we have to do to give `PlatformViewsController` an `Activity`.

xster commented:
Can you reference a bug for remaining work in the platform view theme?

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:
```diff
@@ -66,7 +67,10 @@
-public class FlutterActivity extends FragmentActivity implements OnFirstFrameRenderedListener {
+public class FlutterActivity extends FragmentActivity
+    implements FlutterFragment.FlutterEngineProvider,
+    FlutterFragment.FlutterEngineConfigurator,
```
I'm reading the PR sequentially so I'm making comments based on context from what's above the comment.

I'm getting the impression we were going to just push in and take away an activity from the platformviewcontroller. In that case, why is the activity involved? Couldn't the fragment just always get the activity on its own lifecycle events?

matthew-carroll replied:
I'm not sure what you're asking here. Can you rephrase?

xster replied:
You were saying that this whole file's diff is unrelated to the platform view thing so disregard my comment.

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterActivity.java:
```diff
@@ -353,6 +357,25 @@ protected boolean shouldAttachEngineToActivity() {
+   * Hook for subclasses to easily provide a custom {@code FlutterEngine}.
+   */
+  @Nullable
+  @Override
```
Not clear what's the relationship with these and getFlutterEngine.

matthew-carroll replied:
`getFlutterEngine` has been renamed to `provideFlutterEngine` and its purpose remains unchanged.

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:
```diff
@@ -296,6 +297,8 @@ protected Bundle createArgs() {
   @Nullable
   private PlatformPlugin platformPlugin;
+  @NonNull
+  private final PlatformViewsController platformViewsController = new PlatformViewsController();
```
Are we moving the ownership of the platformviewscontroller out of the plugins registry into here?

matthew-carroll replied:
Yes. As mentioned in the doc for the new plugin registry, its job is now exclusively to register plugins, not to initialize and hold arbitrary resources and dependencies.

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:
```diff
@@ -358,8 +361,14 @@ public void onAttach(Context context) {
+      flutterEngine.getActivityControlSurface().attachToActivity(
+          getActivity(),
+          getLifecycle(),
+          platformViewsController
```
if you own it, you should operate it too
e.g.

platformViewController.here'sTheActivity(getActivity())

matthew-carroll replied:
I don't think that makes much sense. This is typical dependency injection. Notice that `FlutterFragment` doesn't "operate" its `Activity` or `Lifecycle`, it passes those as references to the things that actually work with them. `PlatformViewsController` is no different.

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:
```diff
@@ -398,11 +407,11 @@ private void setupFlutterEngine() {
       // Defer to the Activity that owns us to provide a FlutterEngine.
       Log.d(TAG, "Deferring to attached Activity to provide a FlutterEngine.");
       FlutterEngineProvider flutterEngineProvider = (FlutterEngineProvider) attachedActivity;
-      flutterEngine = flutterEngineProvider.getFlutterEngine(getContext());
```
is this just an API swap?

matthew-carroll replied:
Yeah. I realized in this PR that there were 2 very similar methods called `getFlutterEngine` so I wanted to differentiate.

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:
```diff
@@ -431,10 +440,33 @@ protected FlutterEngine createFlutterEngine(@NonNull Context context) {
+   * subclass needs to override the {@code FragmentActivity}'s behavior, or add to it.
+   */
+  protected void configureFlutterEngine(@NonNull FlutterEngine flutterEngine) {
+    FragmentActivity attachedActivity = getActivity();
```
So this is an auxiliary API to give more control to the activity in case the fragment wants to give the activity to a platfromviewcontroller and the activity says no?

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:
```diff
@@ -538,7 +570,7 @@ public void onStart() {
       @Override
       public void run() {
-        flutterView.attachToFlutterEngine(flutterEngine);
+        flutterView.attachToFlutterEngine(flutterEngine, platformViewsController);
```
I'm surprised something needed to be done to the platform view controller multiple times

matthew-carroll replied:
I'm not sure what you're referring to?

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:
```diff
@@ -783,6 +815,27 @@ protected void onFirstFrameRendered() {}
      * to provide its own {@code FlutterEngine} instance.
```
comment for line 811

> `Returns the {@link FlutterEngine} that should be used by a child {@code FlutterFragment}.` 

This is just an interface right? We don't know that it has a child per se.

matthew-carroll replied:
It's an interface that is defined within `FlutterFragment` for use by `Activity`s that contain `FlutterFragment`s.

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterFragment.java:
```diff
@@ -783,6 +815,27 @@ protected void onFirstFrameRendered() {}
+  }
+
+  /**
+   * Configures a {@link FlutterEngine} after it is created, e.g., adds plugins.
```
Based on this description, it makes me assume that the default FlutterActivity would do something here but it just no-ops.

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterView.java:
```diff
@@ -69,6 +71,8 @@
   private FlutterRenderer.RenderSurface renderSurface;
   private boolean didRenderFirstFrame;
+  @Nullable
+  private PlatformViewsController platformViewsController;
```
I understand why one is given to the flutterview on attachToFlutterEngine but I'm not sure I understand why the flutterview instance then holds onto it.

matthew-carroll replied:
How would we call `detachAccessibilityBridge()` without a reference to the `PlatformViewsController`?

matthew-carroll replied:
Talked offline, resolving old comments.

xster commented on shell/platform/android/io/flutter/embedding/engine/FlutterEnginePluginRegistry.java:
```diff
@@ -259,16 +260,32 @@ private boolean isAttachedToActivity() {
+    // to use it, otherwise an error strack trace will appear that says there is no
+    // flutter/platform_views channel.
+    platformViewsController.attach(
+        activity,
```
So this part I understand. I'm not sure why anyone besides the fragment is involved in FlutterEngineActivityPluginRegistry.attachToActivity.

matthew-carroll replied:
Can you specify exactly what you mean? The reference to "anyone besides the fragment" leaves this ambiguous as to what you're getting at.

matthew-carroll replied:
Talked offline, resolving old comments.

matthew-carroll commented:
> Can you reference a bug for remaining work in the platform view theme?

I'm not sure what this is referring to?

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterView.java:
```diff
@@ -511,7 +528,10 @@ private void resetWillNotDraw(boolean isAccessibilityEnabled, boolean isTouchExp
-  public void attachToFlutterEngine(@NonNull FlutterEngine flutterEngine) {
+  public void attachToFlutterEngine(
+      @NonNull FlutterEngine flutterEngine,
+      @NonNull PlatformViewsController platformViewsController
```
So now if I am just integrating via a FlutterView, I have to create a PlatformViewController too?

matthew-carroll replied:
Yeah, with the way `PlatformViewsController` is integrated with the embedding right now, it's a requirement. If we'd like to make it optional then we'd need to make that change within the platform views subsystem, itself, because right now I'm just matching how platform views integrates with the old embedding.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterView.java:
```diff
@@ -537,7 +558,7 @@ public void attachToFlutterEngine(@NonNull FlutterEngine flutterEngine) {
         this,
         this.flutterEngine.getDartExecutor(),
-        null
+        platformViewsController
```
I see. Can we flip it around? Such that the FlutterView creates and owns the PlatformViewController and since the FlutterFragment only needs this in one place, it does 

```java
flutterEngine.getActivityControlSurface().attachToActivity(
          getActivity(),
          getLifecycle(),
          flutterView.getPlatformViewsController(),
)
```

Regarding the sequencing of onAttach->onCreate->onCreateView etc, I think we might want to tweak it a bit too. Otherwise it's a bit leaky that we're given an Activity which hasn't yet finished initializing to an activity aware plugin. It's a bit like giving a reference to a partially constructed object to someone else in Android equivalent.

If the flutterEngine.getActivityControlSurface().attachToActivity() was on onActivityCreated() instead, then it's both cleaner and onCreateView would have been called by then, meaning you can get the PlatformViewController from the FlutterView.

matthew-carroll replied:
I'll attempt to make this change and see how it goes.

matthew-carroll replied:
Actually, @xster, now I remember why the `PlatformViewsController` isn't created inside `FlutterView` - platform view functionality needs to outlive a given `FlutterView`, right? When the engine is detached from a view, we probably won't attempt to layout or render anything, but we probably can't just tear out the platform views subsystem just because we stopped rendering to the screen, can we?

xster replied:
Doesn't this apply to the FlutterFragment (to which you're giving ownership of the platform view controller in this PR) as well?

xster replied:
Also, I think moving flutterEngine.getActivityControlSurface().attachToActivity() from onAttach to onCreateView is onActivityCreated is a bit orthogonal. I just noticed it now but it wasn't really part of this PR. We should make it cleaner nevertheless either in this PR or in another one.

matthew-carroll replied:
@xster I just pushed a change that moves `PlatformViewsController` into `FlutterEngine` as a component. I think that conceptually makes sense since we've said that some notion of platform views needs to function even in the background. Let me know what you think.

WRT to `onActivityCreated`, we should chat about it. I looked up the Fragment lifecycle and I'm not convinced that there is a symmetric call. I think the question comes down to this: if I attach an existing Fragment to an existing Activity, will we get a call to `onActivityCreated`? If not, then I don't think we can use it.

matthew-carroll commented:
Rebased to resolve conflicts and force pushed. Also added commit that implements a method needed for keyboard support. Verified that accessibility works the same way between the old and new embedding in the webview example project.

xster commented on shell/platform/android/io/flutter/embedding/android/FlutterView.java:
```diff
@@ -511,7 +541,9 @@ private void resetWillNotDraw(boolean isAccessibilityEnabled, boolean isTouchExp
    * {@link FlutterEngine}.
    */
-  public void attachToFlutterEngine(@NonNull FlutterEngine flutterEngine) {
+  public void attachToFlutterEngine(
```
This whole thing looks a lot cleaner now. Thanks!
